pulp_rpm/app/migrations/0016_distribution_tree_one_per_repo.py

```diff
@@ -0,0 +1,44 @@
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+def make_distribution_tree_unique_per_repo(apps, schema_editor):
+    """
+    Ensure there is only one distribution tree per repo and set a repository_id of a repo it belongs to.
+        - if a distribution tree belongs to one repo, just set repository_id
+        - if a distribution tree belongs to multiple repos:
+            - set repository_id for one repo
+            - create a copy of distribution tree, assign it to the next repo, copy all RPMs to the next repo as well
```

Comment:
Do I understand correctly that currently it's possible to have one distribution tree belonging to multiple repos?
Do you agree with what needs to be done in the migration?

Reply:
I believe it cannot belong to multiple repos (not sure),
and I agree with the migration

Reply:
I think it's allowed to copy a distribution tree, it will belong to a different repo but it will still refer the original repository in the variant for the main repo.

Reply:
Are you planning on handling the case where a repo has multiple distribution trees (https://pulp.plan.io/issues/7115)? I'm wondering about how this will be migrated.

Reply:
@daviddavis , this is a hard one. I'm open for suggestions. The question is which one to choose.
I also suspect that it's likely only in theory 2 distribution trees can be there, I doubt there is a real user with such issue there. (I bet tomorrow one will appear after such statement :D).
Maybe it's the case we should ask user to choose and they need to remove one? I mean, I'm not sure even how publish will handle such case and it will be one treeinfo anyway. I guess worth trying to see user experience here.

I can adjust resolve_distribution_trees, so they remove _all_ old ones when the new one is added. But if no new distribution tree is added, I can't choose which one to remove.
Thoughts?

There are 3 option for the migration that I can see:
1. 2 distribution trees can point to the same main repo and the main repo will have RPMs from both :/ Downside: might be a mess and/or dependency conflicts with combining two sets of RPMs.
2. Keep only a distribution tree which points to the main repo a distribution tree is a part of,  and not to any other main repo (aka remove distribution trees which were copied and not synced into a repo).  Downside: it doesn't cover the case when both distribution trees were copied from other repos.
3. Fail migration if there are 2 distribution trees in one repo, and ask user to remove one of the distribution trees from a specific repo. Downside: might be bad upgrade experience.
Any other options?

Reply:
I lean towards 3 as I think it's unlikely there's actual Pulp 3 instance out there that has a repo with two distribution trees.

Reply:
Sounds good to me. Any other opinions since it affects users? @dkliban , @fao89 , @dralley 

@dkliban, if  migraiton fails, can user just easily re-run the installer after a repo is fixed?

If a version user is upgrading to has more than 1 migration (pulp_rpm 3.5 case), and some finished fine but then this one fails, does anyone now if all migrations are rolled back in such case?

Reply:
@daviddavis , I got this error when trying to have two distribution trees (using sync) in one repo centos8 and centos7.1:
`ValueError: Cannot create repository version. Path is duplicated: images/pxeboot/vmlinuz.`

With copy, I got a simialr one, just for different image.
`ValueError: Cannot create repository version. Path is duplicated: images/install.img.`

I'm now even more in favour of option 3 since it's very likely that path overlap prevented two distributions to be in one repo.